extensions_sql_server_virtual_machines.tf

```diff
@@ -0,0 +1,75 @@
+locals {
+  log_analytics_policies = {
+    mdc-log-analytics-arc1-autoprovisioning = {
+      definition_display_name = "[Preview]: Configure Azure Arc-enabled Windows machines with Log Analytics agents connected to default Log Analytics workspace"
+    }
+    mdc-log-analytics-arc2-autoprovisioning = {
+      definition_display_name = "[Preview]: Configure Azure Arc-enabled Linux machines with Log Analytics agents connected to default Log Analytics workspace"
+    }
+  }
+  log_analytics_roles = {
+    role-1 = {
+      name   = "Contributor"
+      policy = "mdc-log-analytics-arc1-autoprovisioning"
+    }
+    role-2 = {
+      name   = "Contributor"
+      policy = "mdc-log-analytics-arc2-autoprovisioning"
+    }
+  }
+}
+
+# Enabling extension - Log Analytics for arc
+data "azurerm_policy_definition" "la_policies" {
+  for_each = (contains(var.mdc_plans_list, "SqlServerVirtualMachines") || contains(var.mdc_plans_list, "Databases")) && !contains(var.mdc_plans_list, "VirtualMachines") ? local.log_analytics_policies : {}
+
+  display_name = each.value.definition_display_name
+}
+
+resource "azurerm_subscription_policy_assignment" "sql" {
+  for_each = (contains(var.mdc_plans_list, "SqlServerVirtualMachines") || contains(var.mdc_plans_list, "Databases")) && !contains(var.mdc_plans_list, "VirtualMachines") ? local.log_analytics_policies : {}
+
+  name                 = each.key
+  policy_definition_id = data.azurerm_policy_definition.la_policies[each.key].id
+  subscription_id      = data.azurerm_subscription.current.id
+  display_name         = each.value.definition_display_name
+  location             = var.location
+
+  identity {
+    type = "SystemAssigned"
+  }
+
+  depends_on = [
+    azurerm_security_center_subscription_pricing.asc_plans["SqlServerVirtualMachines"]
```

Comment:
According to `for_each` expression:

```hcl
contains(var.mdc_plans_list, "SqlServerVirtualMachines") || contains(var.mdc_plans_list, "Databases")) && !contains(var.mdc_plans_list, "VirtualMachines")
```

`azurerm_security_center_subscription_pricing.asc_plans`'s `for_each` expression is:

```hcl
final_plans_list = contains(var.mdc_plans_list, "Databases") ? setsubtract(setunion(var.mdc_plans_list, var.mdc_databases_plans), ["Databases"]) : var.mdc_plans_list
```

What if`mdc_plans_list` contains `Databases` but no `VirtualMachines`, will `azurerm_security_center_subscription_pricing.asc_plans["SqlServerVirtualMachines"]` throw an error?

Reply:
The expression final_plans_list never contain `Databases`. In the expression we check if `Databases` exist and if it is we add 4 different plans (var.mdc_databases_plans) and remove `Databases` from the final_plans_list.
Therefore, instead of `Databases` we always have `SqlServerVirtualMachines`.

The resource azurerm_security_center_subscription_pricing.asc_plans run over final_plans_list so it won't throw an error.

Reply:
Hi @oribenartzyM thanks for the justification, but still exist got a potential issue.

```hcl
 final_plans_list = contains(var.mdc_plans_list, "Databases") ? setsubtract(setunion(var.mdc_plans_list, var.mdc_databases_plans), ["Databases"]) : var.mdc_plans_list
```

When `var.mdc_plans_list` contains `Database`, you'll remove `Databases`, and add all elementes in `var.mdc_databases_plans` to `final_plans_list`, but as this `var.mdc_databases_plans` is an input variable, the user could just override it without `SqlServerVirtualMachines`. In that case we might got an error.

I would recommend let's change `local.enable_sql_server_virtual_machines` to:

```hcl
enable_sql_server_virtual_machines = contains(local.final_plans_list, "SqlServerVirtualMachines") && !contains(var.mdc_plans_list, "VirtualMachines")
```

In that case we could be sure that `azurerm_security_center_subscription_pricing.asc_plans["SqlServerVirtualMachines"]` must be exist.